llvm/lib/Transforms/Vectorize/VPlanTransforms.h

```diff
@@ -163,6 +163,10 @@ struct VPlanTransforms {
   /// Lower abstract recipes to concrete ones, that can be codegen'd.
   static void convertToConcreteRecipes(VPlan &Plan);
 
+  /// Perform instcombine-like simplifications on recipes in \p Plan. Use \p
+  /// CanonicalIVTy as type for all un-typed live-ins in VPTypeAnalysis.
+  static void simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy);
```

Comment:
Would be better for all live-ins to carry their type, e.g., by holding an underlying Undef to hold the Type for live-ins whose Value is not (yet) available, rather than supplying this type externally to VPTypeAnalysis, simplifyRecipes(), etc. In general, better to encode all IR information inside VPlan than to pass it as additional operands to VPlanTransforms.